components/board/genericlinux/gpio.go

```diff
@@ -79,24 +99,89 @@ func (pin *gpioPin) Get(ctx context.Context, extra map[string]interface{}) (resu
 	return (value != 0), nil
 }
 
+// Lock the mutex before calling this! We'll spin up a background goroutine to create a PWM signal
+// in software, if one isn't already running.
+func (pin *gpioPin) startSoftwarePWM() {
+	if pin.pwmRunning {
+		// We're already running a software PWM loop, we don't need another.
+		return
+	}
+	pin.pwmRunning = true
+	pin.waitGroup.Add(1)
+	goutils.ManagedGo(pin.softwarePwmLoop, pin.waitGroup.Done)
+}
+
+// We turn the pin either on or off, and then wait until it's time to turn it off or on again (or
+// until we're supposed to shut down). We return whether we should continue the software PWM cycle.
+func (pin *gpioPin) halfPwmCycle(shouldBeOn bool) bool {
+	pin.mu.Lock()
+	// Before we modify the pin, check if we should stop running
+	if !pin.pwmRunning {
+		pin.mu.Unlock()
+		return false
+	}
+
+	dutyCycle := pin.pwmDutyCyclePct
+	if !shouldBeOn {
+		dutyCycle = 1 - dutyCycle
+	}
+	duration := time.Duration(float64(time.Second) * dutyCycle / float64(pin.pwmFreqHz))
+
+	pin.setInternal(shouldBeOn)
+	pin.mu.Unlock()
+
+	return goutils.SelectContextOrWait(pin.cancelCtx, duration)
+}
+
+func (pin *gpioPin) softwarePwmLoop() {
+	for {
+		if !pin.halfPwmCycle(true) {
+			return
+		}
+		if !pin.halfPwmCycle(false) {
+			return
+		}
+	}
+}
+
 // This helps implement the board.GPIOPin interface for gpioPin.
 func (pin *gpioPin) PWM(ctx context.Context, extra map[string]interface{}) (float64, error) {
-	return math.NaN(), errors.New("PWM stuff is not supported on ioctl GPIO pins yet")
+	pin.mu.Lock()
```

Comment:
Tell me more. I had expected that the RWMutex is much less performant than a standard Mutex, and is only preferred if you've got multiple long-lived read-only sections (e.g., you're building a database). Quoting [a description of how RWMutex works](https://sreramk.medium.com/go-sync-rwmutex-internals-and-usage-explained-9eb15865bba): 
> Drawbacks of using sync.RWMutex:
>The [atomic operations performed on integers] in `RLock()` and `RUnlock()` respectively causes cache-contention which degrades performance, with frequent calls to them.

and quoting an [obsolete ticket complaining of RWLock performance](https://github.com/golang/go/issues/38813):
> [`RWMutex` is] pretty much sure to be slower than `Mutex` since it has a more complicated job to do.

If there's a reason to think `RWMutex` is at least as performant as normal `Mutex`, I'm happy to make this change, but I don't see the benefit yet.

Reply:
fair point, we should leave as is